File: backend/app/services/ai/quiz.py
"""
Quiz Generator — Generates MCQ quizzes and analyzes knowledge gaps via Gemini.
"""
from app.db.models import Topic, Subtopic, KnowledgeGap
from sqlalchemy.orm import Session
import json
from app.services.ai import get_gemini_response


def generate_quiz(topic_id: int, user_id: int, db: Session) -> dict:
    """Generate a quiz for a topic, difficulty adapts to topic status."""
    topic = db.query(Topic).filter(Topic.id == topic_id).first()
    if not topic:
        return {"questions": []}

    subtopics = db.query(Subtopic).filter(Subtopic.topic_id == topic_id).all()
    context = "\n\n".join([f"## {s.title}\n{s.content}" for s in subtopics if s.content])

    # Adaptive difficulty
    difficulty = "medium"
    if topic.status == "Strong":
        difficulty = "hard"
    elif topic.status == "Weak":
        difficulty = "easy"

    base_context = f'Base your questions on this study material:\n{context[:4000]}' if context else f'Create questions about "{topic.title}" using general knowledge.'

    prompt = f"""You are an AI teacher. Generate a quiz with 5 multiple-choice questions about "{topic.title}".
Difficulty level: {difficulty}

{base_context}

Return ONLY a raw JSON object (no markdown backticks, no extra text) matching this exact schema:
{{
  "questions": [
    {{
      "text": "Question text here?",
      "options": ["Option A", "Option B", "Option C", "Option D"],
      "correct_answer_index": 0,
      "explanation": "Brief explanation of why this answer is correct"
    }}
  ]
}}"""

    text = get_gemini_response(prompt)
    try:
        text = text.strip()
        if text.startswith('```json'):
            text = text[7:]
        if text.startswith('```'):
            text = text[3:]
        if text.endswith('```'):
            text = text[:-3]
        return json.loads(text.strip())
    except Exception:
        return {"questions": []}


def analyze_gaps(topic_id: int, score: int, total_questions: int, user_id: int, db: Session):
    """Analyze quiz performance, update topic status, detect knowledge gaps."""
    topic = db.query(Topic).filter(Topic.id == topic_id).first()
    if not topic:
        return

    percentage = (score / total_questions * 100) if total_questions > 0 else 0

    if percentage >= 80:
        topic.status = "Strong"
    elif percentage >= 50:
        topic.status = "Medium"
    else:
        topic.status = "Weak"
        # Add knowledge gap if not already tracked
        existing = db.query(KnowledgeGap).filter(
            KnowledgeGap.user_id == user_id,
            KnowledgeGap.suggested_topic == topic.title
        ).first()
        if not existing:
            gap = KnowledgeGap(
                user_id=user_id,
                description=f"Low score ({score}/{total_questions}) on {topic.title}. Consider revising this topic and its subtopics.",
                suggested_topic=topic.title
            )
            db.add(gap)

    db.commit()

File: server/routes/quiz.js
const router = require('express').Router();
const auth   = require('../middleware/auth');
const Topic  = require('../models/Topic');
const { QuizAttempt, KnowledgeGap } = require('../models/Quiz');
const { gemini } = require('../services/ai');

// POST /api/quiz/generate
router.post('/generate', auth, async (req, res) => {
  try {
    const { topic_id } = req.body;
    const topic = await Topic.findOne({ _id: topic_id, user: req.user._id });
    if (!topic) return res.status(404).json({ message: 'Topic not found' });

    const difficulty = topic.status === 'Weak' ? 'challenging' : topic.status === 'Strong' ? 'advanced' : 'moderate';
    const subtopicsText = topic.subtopics.map(s => `- ${s.title}: ${s.content}`).join('\n');

    const prompt = `Generate exactly 5 ${difficulty} multiple choice questions to test understanding of "${topic.title}".
Context:
${subtopicsText}

Return ONLY a JSON array (no markdown):
[{"question":"...","options":["A) ...","B) ...","C) ...","D) ..."],"correct":0,"explanation":"..."}]`;

    const raw   = await gemini(prompt, req.user.geminiApiKey);
    const clean = raw.replace(/```json|```/g, '').trim();
    const quiz  = JSON.parse(clean);
    res.json({ topic_id, quiz });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// POST /api/quiz/submit
router.post('/submit', auth, async (req, res) => {
  try {
    const { topic_id, score, total_questions } = req.body;
    await QuizAttempt.create({ user: req.user._id, topic: topic_id, score, totalQuestions: total_questions });

    const pct = total_questions > 0 ? (score / total_questions) * 100 : 0;
    const status = pct >= 80 ? 'Strong' : pct >= 50 ? 'Medium' : 'Weak';
    await Topic.updateOne({ _id: topic_id }, { status });

    if (pct < 60) {
      const topic = await Topic.findById(topic_id);
      await KnowledgeGap.create({
        user: req.user._id,
        description: `Low quiz score (${pct.toFixed(0)}%) on "${topic?.title}"`,
        suggestedTopic: topic?.title || '',
        source: 'quiz'
      });
    }
    res.json({ message: 'Quiz submitted', new_status: status });
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// GET /api/quiz/gaps
router.get('/gaps', auth, async (req, res) => {
  const gaps = await KnowledgeGap.find({ user: req.user._id }).sort({ createdAt: -1 }).limit(20);
  res.json(gaps);
});

module.exports = router;
